scripts/inspect_train_data.py

```diff
@@ -2,28 +2,121 @@
+from olmo.data import build_memmap_dataset, build_train_dataloader
+from olmo.data.iterable_dataset import IterableDataset
 from olmo.tokenizer import Tokenizer
-from olmo.util import clean_opt, prepare_cli_environment
+from olmo.util import add_cached_path_clients, clean_opt, prepare_cli_environment
+
+
+def get_global_train_examples_seen_before_step(step: int, trainer_state: dict, cfg: TrainConfig):
+    global_step = trainer_state["global_step"]
+
+    if global_step >= step:
+        raise ValueError(f"Step {step} must be after training first step {global_step}")
+
+    global_train_examples_seen_this_epoch = trainer_state.get(
+        "global_train_examples_seen_this_epoch",
+        trainer_state.get(  # for backwards compatibility
+            "global_train_examples_seen",
+            trainer_state.get("global_data_step", global_step) * cfg.global_train_batch_size,
+        ),
+    )
+
+    # Subtract 1 from step because we want to be just before that step
+    global_train_examples_seen_this_epoch += (step - 1 - global_step) * cfg.global_train_batch_size
+    return global_train_examples_seen_this_epoch
 
 
-def main(save_folder: Path, *steps: int, rank: Optional[int] = None):
+def inspect_data_without_device_data_indices(
+    run_path: str, *steps: int, world_size: int, ranks: List[int], reference_step: int
+):
+    cfg = TrainConfig.load(
+        cached_path(os.path.join(run_path, f"step{reference_step}/config.yaml")),
+        overrides=[clean_opt("--evaluators=[]"), clean_opt("--save_overwrite")],
+    )
+    cfg.data.num_workers = 1
+
+    if cfg.global_train_batch_size % world_size != 0:
+        raise ValueError(f"World size must divide global_train_batch_size {cfg.global_train_batch_size}")
+    cfg.device_train_batch_size = cfg.global_train_batch_size // world_size
+
+    try:
+        trainer_state = load_state_dict(run_path, f"step{reference_step}/train/rank0.pt", map_location="cpu")
+    except FileNotFoundError:
+        try:
+            # Unsharded checkpointing
+            trainer_state = load_state_dict(run_path, f"step{reference_step}/train.pt", map_location="cpu")
+        except FileNotFoundError:
+            # Legacy checkpointing
+            trainer_state = load_state_dict(run_path, f"step{reference_step}/rank0.pt", map_location="cpu")
+
+    tokenizer = Tokenizer.from_train_config(cfg)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        cfg.save_folder = tmpdir
+
+        # Build dataloader in rank 0 to generate the indices file
+        os.environ["RANK"] = "0"
+        dataloader = build_train_dataloader(cfg, world_size=world_size)
+
+        for rank in ranks:
+            os.environ["RANK"] = str(rank)
+            # Set FS_LOCAL_RANK to a non-zero number so that global data indices are not rewritten
+            os.environ["FS_LOCAL_RANK"] = "1"
+
+            for step in steps:
+                dataloader = build_train_dataloader(cfg, world_size=world_size)
```

Comment:
This would have to rebuild the indices file every time, right? That could be slow, but we could probably avoid rebuilding for every rank.

Reply:
Setting `FS_LOCAL_RANK=1` avoids rebuilding the indices file every time since it's only down for local FS rank 0.
`# Set FS_LOCAL_RANK to a non-zero number so that global data indices are not rewritten`

Reply:
[d137a1f](https://github.com/allenai/OLMo/pull/593/commits/d137a1fe93785487382564883603d98736bf6d15)